### PR TITLE
OCPBUGS-17788: Improved error handling for missing MC

### PR DESCRIPTION
**Pull request by cdoern**
the error for when a MachineConfig is missing is very vague. Usually the error has little to do with the cause. Specify steps users can take to alleviate the situations.

### CODE REVIEW

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -1999,7 +2003,13 @@ func (dn *Daemon) checkStateOnFirstRun() error {
 	}
 	// currentConfig != desiredConfig, and we're not booting up into the desiredConfig.
 	// Kick off an update.
-	return dn.triggerUpdate(state.currentConfig, state.desiredConfig, state.currentImage, state.desiredImage)
+	err = dn.triggerUpdate(state.currentConfig, state.desiredConfig, state.currentImage, state.desiredImage)
+	if err != nil && apierrors.IsNotFound(err) {
+		// if this is first run, the MC we went to get DNE this means there was a content mismatch
+		// bwtn what user provided and what was rendered. State that here.
+		return fmt.Errorf("User provided configuration caused a mismatch in the rendered config. Please compare /etc/mcs-machine-config-content.json and the existing rendered config. %w", err)
```
Comment (yuqi-zhang) [inline pkg/daemon/daemon.go:2010]: So, this error here actually triggers on the curentConfig, meaning that we don't get to this line if we have this issue. 

I think you would need to add it to almost the beginning of the function, in `getStateAndConfigs` return if you want this. Maybe a better phrasing would be:

```
Could not find config %v in-cluster, this likely indicates the MachineConfigs in-cluster has changed during the install process. If you are seeing this when installing the cluster, please compare the in-cluster rendered machineconfigs to ...
```

or is that a bit too wordy? You are theoretically able to hit this later on if you delete a config while the node is rebooting, for example, so I don't want to confuse the user

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -792,6 +792,10 @@ func (dn *Daemon) syncNode(key string) error {
 		}
 
 		if err := dn.triggerUpdate(ufc.currentConfig, ufc.desiredConfig, ufc.currentImage, ufc.desiredImage); err != nil {
+			if apierrors.IsNotFound(err) {
+				// if MC was not found, let user know where they can find more info on this.
+				return fmt.Errorf("A rendered MC was removed. Remove all currentConfig or desiredConfig annotations from the nodes using this MC. Also delete /etc/machine-config-daemon/currentconfig: %w ", err)
```
Comment (yuqi-zhang) [inline pkg/daemon/daemon.go:797]: Hmm, this (and the next one) feels less like error messages and more like "what to do in case of this error".

The phrasing sounds like we want the user to delete the annotation entirely (which I think will break the node as well?) but telling them to switch to another rendered config might just get them into a worse spot if they are uncertain what they are doing.

It's probably best to either somehow use an alert->runbook method to self help or maybe just clarify what the error is here. It's better that they can reach out to support with the issue (a MachineConfig seems to have been deleted)

Also I feel like maybe we'd hit this more in the `prepUpdateFromCluster` above or other locations and not in this function? I feel like the timing would be very tight

**Comment by cdoern** [comment]
@yuqi-zhang i'll convert it to an alert

**Comment by cdoern** [comment]
@yuqi-zhang can you PTAL at this? converted it to an alert

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml [hunk 3]
```diff
@@ -160,3 +160,16 @@ spec:
...
+          expr: |
+            mcd_missing_mc > 0
+          labels:
+            namespace: openshift-machine-config-operator
+            severity: warning
+          annotations:
+            summary: "This keeps track of Machine Config failures. Specifically a common failure on install when a rendered Machine Config is missing. Triggered when this error happens once."
+            description: >-
+              Could not find config {{ $labels.mc }} in-cluster, this likely indicates the MachineConfigs in-cluster has changed during the install process. 
+              If you are seeing this when installing the cluster, please compare the in-cluster rendered machineconfigs to s/etc/mcs-machine-config-content.json
```
Comment (yuqi-zhang) [inline install/0000_90_machine-config-operator_01_prometheus-rules.yaml:175]: Curious, for the extra s in `s/etc/mcs-machine-config-content.json`, is that for formatting?

File: pkg/daemon/daemon.go [hunk 4]
```diff
@@ -791,7 +795,11 @@ func (dn *Daemon) syncNode(key string) error {
 			return err
 		}
 
-		if err := dn.triggerUpdate(ufc.currentConfig, ufc.desiredConfig, ufc.currentImage, ufc.desiredImage); err != nil {
+		if missingMC, err := dn.triggerUpdate(ufc.currentConfig, ufc.desiredConfig, ufc.currentImage, ufc.desiredImage); err != nil {
+			// if MC was not found, let user know where they can find more info on this.
+			if apierrors.IsNotFound(err) {
+				mcdMissingMC.WithLabelValues(missingMC).Inc()
```
Comment (yuqi-zhang) [inline pkg/daemon/daemon.go:801]: This is using the same current/desired config from the above `dn.prepUpdateFromCluster()` call right? If MC was missing, wouldn't the previous `Inc()` already cover this?

File: install/0000_90_machine-config-operator_01_prometheus-rules.yaml (hunk 3 above)
Comment (cdoern) [reply install/0000_90_machine-config-operator_01_prometheus-rules.yaml:175]: fixed

File: pkg/daemon/daemon.go (hunk 4 above)
Comment (cdoern) [reply pkg/daemon/daemon.go:801]: i added this one in for safety. I think there is a path where we can get to this and not the other. All that matters is that they both can't happen!

File: pkg/daemon/daemon.go [hunk 5]
```diff
@@ -1900,8 +1908,12 @@ func (dn *Daemon) checkStateOnFirstRun() error {
 	// Update our cached copy
 	dn.node = node
 
+	currentConfigName, _ := getNodeAnnotation(dn.node, constants.CurrentMachineConfigAnnotationKey)
```
[the remark this answers is not in the document]
Comment (cdoern) [reply pkg/daemon/daemon.go:1911]: true. I will make it so that we only inc if the error to get the annotation is nil.

**Comment by cheesesashimi** [comment]
Overall, this looks good. I have some reservations about only returning a specific value from `dn.triggerUpdate()` (and everything else in that call chain) when an error has occurred. Instead, it would be more idiomatic to put the name of the missing MachineConfig into a custom error type and extract it when needed. Here's a way to do this:

```golang
// First, we create a custom error type to hold the missing MachineConfig name.
type ErrMissingMachineConfig struct {
	missingMC string
}

// Optional constructor.
func newErrMissingMachineConfig(missingMC string) error {
	return &ErrMissingMachineConfig{
		missingMC: missingMC,
	}
}

// This implements the error interface within Go.
func (e *ErrMissingMachineConfig) Error() string {
	return fmt.Sprintf("missing MachineConfig %s", e.missingMC)
}

// This is an optional accessor to get the missing MachineConfig.
func (e *ErrMissingMachineConfig) MissingMachineConfig() string {
	return e.missingMC
}
```

Next, within `triggerUpdate()` (and everything in that call path), we do something like this instead:
```golang
func (dn *Daemon) triggerUpdate() error {
	// ...
	desiredConfig, err = dn.mcLister.Get(dcAnnotation)
	if err != nil {
		// errors.Join() is a new built-in, see: https://pkg.go.dev/errors#example-Join
		// It has a similar purpose to the AggregateError type found here: https://github.com/kubernetes/apimachinery/blob/master/pkg/util/errors/errors.go#L41-L46
		return errors.Join(newErrMissingMachineConfig(dcAnnotation), err)
	}
	// ...
}
```

Finally, to determine whether we have an `ErrMissingMachineConfig`, we can examine the error chain like this:
```golang
func caller() {
	err := dn.triggerUpdate()
	if err != nil {
		missingMCErr := &ErrMissingMachineConfig{}
		// Here, we can check if we have an ErrMissingMachineConfig error, extract
		// the missing MachineConfig from it, and report it. All without needing to
		// add an additional parameter onto everything :).
		if errors.As(err, *missingMCErr) {
			mcdMissingMC.WithLabelValues(missingMCErr.MissingMachineConfig()).Inc()
		}
	}

	// ...
}
```

**Review by cheesesashimi — CHANGES_REQUESTED** [review]
Overall, this looks good and we're almost there! I have a few additional suggestions in the meantime.
